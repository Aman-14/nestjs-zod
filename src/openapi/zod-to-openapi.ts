import { Type } from '@nestjs/common'
import { SchemaObject } from '@nestjs/swagger/dist/interfaces/open-api-spec.interface'
import mergeDeep from 'merge-deep'
import {
  ZodArray,
  ZodBigInt,
  ZodBoolean,
  ZodDateString,
  ZodDefault,
  ZodEnum,
  ZodLiteral,
  ZodNullable,
  ZodNumber,
  ZodObject,
  ZodOptional,
  ZodRecord,
  ZodSet,
  ZodString,
  ZodTransformer,
  ZodTuple,
  ZodTypeAny,
  ZodUnion,
  ZodIntersection,
  ZodDiscriminatedUnion,
} from '../z'

export function is<T extends Type<ZodTypeAny>>(
  input: ZodTypeAny,
  factory: T
): input is InstanceType<T> {
  return factory.name === input.constructor.name
}
export function zodToOpenAPI(zodType: ZodTypeAny) {
  const object: SchemaObject = {}

  if (zodType.description) {
    object.description = zodType.description
  }

  if (is(zodType, ZodString)) {
    const { checks } = zodType._def
    object.type = 'string'

    for (const check of checks) {
      if (check.kind === 'min') {
        object.minLength = check.value
      } else if (check.kind === 'max') {
        object.maxLength = check.value
      } else if (check.kind === 'email') {
        object.format = 'email'
      } else if (check.kind === 'url') {
        object.format = 'uri'
      } else if (check.kind === 'uuid') {
        object.format = 'uuid'
      } else if (check.kind === 'cuid') {
        object.format = 'cuid'
      } else if (check.kind === 'regex') {
        object.pattern = check.regex.source
      }
    }
  }

  if (is(zodType, ZodBoolean)) {
    object.type = 'boolean'
  }

  if (is(zodType, ZodNumber)) {
    const { checks } = zodType._def
    object.type = 'number'

    for (const check of checks) {
      if (check.kind === 'int') {
        object.type = 'integer'
      } else if (check.kind === 'min') {
        object.minimum = check.value
        object.exclusiveMinimum = !check.inclusive
      } else if (check.kind === 'max') {
        object.maximum = check.value
        object.exclusiveMaximum = !check.inclusive
      } else if (check.kind === 'multipleOf') {
        object.multipleOf = check.value
      }
    }
  }

  if (is(zodType, ZodDateString)) {
    const { checks } = zodType._def
    object.type = 'string'

    for (const check of checks) {
      if (check.kind === 'format') {
        object.format = check.value
      }
    }
  }

  if (is(zodType, ZodBigInt)) {
    object.type = 'integer'
    object.format = 'int64'
  }

  if (is(zodType, ZodArray)) {
    const { minLength, maxLength, type } = zodType._def
    object.type = 'array'
    if (minLength) object.minItems = minLength.value
    if (maxLength) object.maxItems = maxLength.value
    object.items = zodToOpenAPI(type)
  }

  if (is(zodType, ZodTuple)) {
    const { items } = zodType._def
    object.type = 'array'
    object.items = { oneOf: items.map(zodToOpenAPI) }
  }

  if (is(zodType, ZodSet)) {
    const { valueType, minSize, maxSize } = zodType._def
    object.type = 'array'
    if (minSize) object.minItems = minSize.value
    if (maxSize) object.maxItems = maxSize.value
    object.items = zodToOpenAPI(valueType)
    object.uniqueItems = true
  }

  if (is(zodType, ZodUnion)) {
    const { options } = zodType._def
    object.oneOf = options.map(zodToOpenAPI)
  }

  if (is(zodType, ZodDiscriminatedUnion)) {
    const { options } = zodType._def
    object.oneOf = []
    for (const schema of options.values()) {
      object.oneOf.push(zodToOpenAPI(schema))
    }
  }

  if (is(zodType, ZodLiteral)) {
    const { value } = zodType._def

    if (typeof value === 'string') {
      object.type = 'string'
      object.enum = [value]
    }

    if (typeof value === 'number') {
      object.type = 'number'
      object.minimum = value
      object.maximum = value
    }

    if (typeof value === 'boolean') {
      // currently there is no way to completely describe boolean literal
      object.type = 'boolean'
    }
  }

  if (is(zodType, ZodEnum)) {
    const { values } = zodType._def
    object.type = 'string'
    object.enum = values
  }

  if (is(zodType, ZodTransformer)) {
    const { schema } = zodType._def
    Object.assign(object, zodToOpenAPI(schema))
  }

  if (is(zodType, ZodNullable)) {
    const { innerType } = zodType._def
    Object.assign(object, zodToOpenAPI(innerType))
    object.nullable = true
  }

  if (is(zodType, ZodOptional)) {
    const { innerType } = zodType._def
    Object.assign(object, zodToOpenAPI(innerType))
  }

  if (is(zodType, ZodDefault)) {
    const { defaultValue, innerType } = zodType._def
    Object.assign(object, zodToOpenAPI(innerType))
    object.default = defaultValue()
  }

  if (is(zodType, ZodObject)) {
    const { shape } = zodType._def
    object.type = 'object'

    object.properties = {}
    object.required = []

    for (const [key, schema] of Object.entries<ZodTypeAny>(shape())) {
      object.properties[key] = zodToOpenAPI(schema)
      const optionalTypes = [ZodOptional.name, ZodDefault.name]
      const isOptional = optionalTypes.includes(schema.constructor.name)
      if (!isOptional) object.required.push(key)
    }
  }

  if (is(zodType, ZodRecord)) {
    const { valueType } = zodType._def
    object.type = 'object'
    object.additionalProperties = zodToOpenAPI(valueType)
  }

  if (is(zodType, ZodIntersection)) {
    const { left, right } = zodType._def
    const merged = mergeDeep(zodToOpenAPI(left), zodToOpenAPI(right))
    Object.assign(object, merged)
  }

  return object
}
