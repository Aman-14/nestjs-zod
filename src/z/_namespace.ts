/*
 * Just for test purposes
 */
export * as z from './index'
