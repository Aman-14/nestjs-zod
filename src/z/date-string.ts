import {
  INVALID,
  ParseInput,
  ParseReturnType,
  ParseStatus,
  ZodParsedType,
  ZodType,
  ZodTypeDef,
} from 'zod'
import { ZodIssueCode, addIssueToContext } from './issues'
import {
  ErrorMessage,
  findCheck,
  hasCheck,
  normalizeErrorMessage,
  processCreateParams,
  RawCreateParams,
} from './shared'
import { ZodFirstPartyTypeKindExtended } from './type-names'

type ZodIsoDateCheck =
  | { kind: 'minYear'; value: number; message?: string }
  | { kind: 'maxYear'; value: number; message?: string }
  | { kind: 'weekDay'; message?: string }
  | { kind: 'weekend'; message?: string }

export interface ZodDateStringDef extends ZodTypeDef {
  checks: ZodIsoDateCheck[]
  typeName: ZodFirstPartyTypeKindExtended
}

export class ZodDateString extends ZodType<string, ZodDateStringDef> {
  _parse(input: ParseInput): ParseReturnType<string> {
    const parsedType = this._getType(input)
    const context = this._getOrReturnCtx(input)

    const notString = parsedType !== ZodParsedType.string
    const date = new Date(input.data)

    if (notString || Number.isNaN(date.getTime())) {
      addIssueToContext(context, {
        code: ZodIssueCode.invalid_type,
        expected: ZodParsedType.string,
        received: context.parsedType,
      })

      return INVALID
    }

    const status = new ParseStatus()

    for (const check of this._def.checks) {
      if (check.kind === 'minYear') {
        const invalid = date.getFullYear() < check.value
        if (!invalid) continue

        addIssueToContext(context, {
          code: ZodIssueCode.too_small,
          type: 'iso_date_year',
          minimum: check.value,
          inclusive: true,
          message: check.message,
        })

        status.dirty()
      } else if (check.kind === 'maxYear') {
        const invalid = date.getFullYear() > check.value
        if (!invalid) continue

        addIssueToContext(context, {
          code: ZodIssueCode.too_big,
          type: 'iso_date_year',
          maximum: check.value,
          inclusive: true,
          message: check.message,
        })

        status.dirty()
      } else if (check.kind === 'weekDay') {
        const day = date.getDay()
        const invalid = day === 0 || day === 6
        if (!invalid) continue

        addIssueToContext(context, {
          code: ZodIssueCode.invalid_date_string_day,
          expected: 'weekDay',
          message: check.message,
        })

        status.dirty()
      } else if (check.kind === 'weekend') {
        const day = date.getDay()
        const invalid = day !== 0 && day !== 6
        if (!invalid) continue

        addIssueToContext(context, {
          code: ZodIssueCode.invalid_date_string_day,
          expected: 'weekend',
          message: check.message,
        })

        status.dirty()
      }
    }

    return { status: status.value, value: input.data }
  }

  _addCheck(check: ZodIsoDateCheck) {
    return new ZodDateString({
      ...this._def,
      checks: [...this._def.checks, check],
    })
  }

  static create = (params?: RawCreateParams): ZodDateString => {
    return new ZodDateString({
      checks: [],
      typeName: ZodFirstPartyTypeKindExtended.ZodDateString,
      ...processCreateParams(params),
    })
  }

  minYear(year: number, message?: ErrorMessage) {
    return this._addCheck({
      kind: 'minYear',
      value: year,
      ...normalizeErrorMessage(message),
    })
  }

  maxYear(year: number, message?: ErrorMessage) {
    return this._addCheck({
      kind: 'maxYear',
      value: year,
      ...normalizeErrorMessage(message),
    })
  }

  weekDay(message?: ErrorMessage) {
    return this._addCheck({
      kind: 'weekDay',
      ...normalizeErrorMessage(message),
    })
  }

  weekend(message?: ErrorMessage) {
    return this._addCheck({
      kind: 'weekend',
      ...normalizeErrorMessage(message),
    })
  }

  get _minYear() {
    return findCheck(this._def.checks, 'minYear')
  }

  get _maxYear() {
    return findCheck(this._def.checks, 'maxYear')
  }

  get isWeekDay() {
    return hasCheck(this._def.checks, 'weekDay')
  }

  get isWeekend() {
    return hasCheck(this._def.checks, 'weekend')
  }
}

export const dateString = ZodDateString.create
