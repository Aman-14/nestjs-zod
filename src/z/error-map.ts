import {
  ZodErrorMap,
  defaultErrorMap,
  ZodIssueOptionalMessage as ZodIssueOptionalMessageDefault,
  setErrorMap,
} from 'zod'
import {
  DateStringFormat,
  ZodIssueCode,
  ZodIssueOptionalMessage,
} from './issues'

type ErrorMapContext = Parameters<ZodErrorMap>[1]

type ZodErrorMapExtended = (
  issue: ZodIssueOptionalMessage,
  context: ErrorMapContext
) => ReturnType<ZodErrorMap>

const extendedErrorMap: ZodErrorMapExtended = (issue, context) => {
  /*
   * At first, we should handle the custom Issues,
   * because defaultErrorMap throws an Error when no match found
   */

  if (issue.code === ZodIssueCode.invalid_date_string_format) {
    const mapper: Record<DateStringFormat, string> = {
      'date': 'YYYY-MM-DD',
      'date-time': 'YYYY-MM-DDTHH:mm:ssZ',
    }

    const readable = mapper[issue.expected]
    const message = `Invalid date, expected it to match format: ${readable}`
    return { message }
  }

  if (issue.code === ZodIssueCode.invalid_date_string_direction) {
    const message = `Invalid date, expected it to be the ${issue.expected}`
    return { message }
  }

  if (issue.code === ZodIssueCode.invalid_date_string_day) {
    const mapper: Record<typeof issue['expected'], string> = {
      weekDay: 'week day',
      weekend: 'weekend',
    }

    const readable = mapper[issue.expected]
    const message = `Invalid date, expected it to be a ${readable}`
    return { message }
  }

  if (
    issue.code === ZodIssueCode.too_small &&
    issue.type === 'date_string_year'
  ) {
    const appendix = issue.inclusive ? 'or equal to ' : ''
    const message = `Year must be greater than ${appendix}${issue.minimum}`
    return { message }
  }

  if (
    issue.code === ZodIssueCode.too_big &&
    issue.type === 'date_string_year'
  ) {
    const appendix = issue.inclusive ? 'or equal to ' : ''
    const message = `Year must be less than ${appendix}${issue.maximum}`
    return { message }
  }

  /*
   * At this moment, there are no custom Issues remained,
   * so we can safely delegate the control to defaultErrorMap
   */

  return defaultErrorMap(issue as ZodIssueOptionalMessageDefault, context)
}

function setExtendedErrorMap(map: ZodErrorMapExtended) {
  setErrorMap(map)
}

// set custom errorMap before any user code is executed
setExtendedErrorMap(extendedErrorMap)

export {
  extendedErrorMap as defaultErrorMap,
  setExtendedErrorMap as setErrorMap,
}
export type { ZodErrorMapExtended as ZodErrorMap }
