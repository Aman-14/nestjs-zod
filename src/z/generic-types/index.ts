export * from './json'
export * from './from'
