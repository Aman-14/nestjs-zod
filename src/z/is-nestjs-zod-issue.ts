import { ZodIssueCode } from 'zod'
import { NestJsZodIssue, ZodIssueOptionalMessage } from './issues'

export function isNestJsZodIssue(
  issue: ZodIssueOptionalMessage
): issue is NestJsZodIssue {
  return issue.code === ZodIssueCode.custom && issue.params?.isNestJsZod
}
