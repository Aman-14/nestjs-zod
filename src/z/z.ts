/* eslint-disable import/export */
export * from 'zod'

export { defaultErrorMap, setErrorMap } from './error-map'
export type { ZodErrorMap } from './error-map'

export { ZodIssueCode, addIssueToContext } from './issues'
export type {
  ZodIssue,
  ZodIssueOptionalMessage,
  ZodInvalidIsoDateDayIssue,
  ZodTooBigIssue,
  ZodTooSmallIssue,
} from './issues'

export * from './type-names'
export * from './date-string'
