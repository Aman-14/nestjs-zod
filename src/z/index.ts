/* eslint-disable import/export */
import * as z from './z'

export * from './z'
export { z }
