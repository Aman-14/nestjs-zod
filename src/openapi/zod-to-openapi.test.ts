import { ZodTypeAny } from '../z'
import { z } from '../z/_namespace'
import { zodToOpenAPI } from './zod-to-openapi'

const complexTestSchema = z.object({
  stringMinMax: z.string().min(5).max(15),
  numberMinMax: z.number().min(3).max(10),
  numberMultipleOf: z.number().multipleOf(10),
  enumDefault: z.enum(['buy-it-now', 'auctions']).default('buy-it-now'),
  tuple: z.tuple([
    z.string(),
    z.object({ name: z.literal('Rudy') }),
    z.array(z.union([z.literal('blue'), z.literal('red')])),
  ]),
  union: z
    .union([z.literal('kek'), z.literal('topkek')])
    .describe('Please choose topkek'),
  literalString: z.literal('123'),
  literalNumber: z.literal(123),
  literalBoolean: z.literal(true),
  array: z.array(z.string()).max(3),
  objectNested: z.object({
    uuid: z.string().uuid(),
  }),
  record: z.record(z.number()),
  recordWithKeys: z.record(z.number(), z.string()),
  dateString: z.dateString().cast().describe('My date string'),
})

it('should serialize a complex schema', () => {
  const openApiObject = zodToOpenAPI(complexTestSchema)

  expect(openApiObject).toMatchSnapshot()
})

it('should serialize objects', () => {
  const schema = z.object({
    prop1: z.string(),
    prop2: z.string().optional(),
  })
  const openApiObject = zodToOpenAPI(schema)

  expect(openApiObject).toEqual({
    type: 'object',
    required: ['prop1'],
    properties: {
      prop1: {
        type: 'string',
      },
      prop2: {
        type: 'string',
      },
    },
  })
})

it('should serialize nullable types', () => {
  const schema = z.string().nullable()
  const openApiObject = zodToOpenAPI(schema)

  expect(openApiObject).toEqual({ type: 'string', nullable: true })
})

it('should serialize optional types', () => {
  const schema = z.string().optional()
  const openApiObject = zodToOpenAPI(schema)

  expect(openApiObject).toEqual({ type: 'string' })
})

it('should serialize types with default value', () => {
  const schema = z.string().default('abitia')
  const openApiObject = zodToOpenAPI(schema)

  expect(openApiObject).toEqual({ type: 'string', default: 'abitia' })
})

it('should serialize enums', () => {
  const schema = z.enum(['adama', 'kota'])
  const openApiObject = zodToOpenAPI(schema)

  expect(openApiObject).toEqual({
    type: 'string',
    enum: ['adama', 'kota'],
  })
})

describe('scalar types', () => {
  const testCases: [ZodTypeAny, string, string?][] = [
    // [zod type, expected open api type, expected format]
    [z.string(), 'string'],
    [z.number(), 'number'],
    [z.boolean(), 'boolean'],
    [z.bigint(), 'integer', 'int64'],
    // [z.null(), 'null'], <- Needs OpenApi 3.1 to be represented correctly
    // [z.undefined(), 'undefined'], <- TBD, probably the property should be removed from schema
  ]

  for (const [zodType, expectedType, expectedFormat] of testCases) {
    // eslint-disable-next-line no-loop-func
    it(expectedType, () => {
      const openApiObject = zodToOpenAPI(zodType)

      expect(openApiObject).toEqual({
        type: expectedType,
        format: expectedFormat ?? undefined,
      })
    })
  }
})
