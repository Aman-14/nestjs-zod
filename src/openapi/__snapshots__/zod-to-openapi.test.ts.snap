// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should serialize a complex schema 1`] = `
Object {
  "properties": Object {
    "array": Object {
      "items": Object {
        "type": "string",
      },
      "maxItems": 3,
      "type": "array",
    },
    "dateString": Object {
      "description": "My date string",
      "format": "date-time",
      "type": "string",
    },
    "enumDefault": Object {
      "default": "buy-it-now",
      "enum": Array [
        "buy-it-now",
        "auctions",
      ],
      "type": "string",
    },
    "literalBoolean": Object {
      "type": "boolean",
    },
    "literalNumber": Object {
      "maximum": 123,
      "minimum": 123,
      "type": "number",
    },
    "literalString": Object {
      "enum": Array [
        "123",
      ],
      "type": "string",
    },
    "numberMinMax": Object {
      "exclusiveMaximum": false,
      "exclusiveMinimum": false,
      "maximum": 10,
      "minimum": 3,
      "type": "number",
    },
    "numberMultipleOf": Object {
      "multipleOf": 10,
      "type": "number",
    },
    "objectNested": Object {
      "properties": Object {
        "uuid": Object {
          "format": "uuid",
          "type": "string",
        },
      },
      "required": Array [
        "uuid",
      ],
      "type": "object",
    },
    "record": Object {
      "additionalProperties": Object {
        "type": "number",
      },
      "type": "object",
    },
    "recordWithKeys": Object {
      "additionalProperties": Object {
        "type": "string",
      },
      "type": "object",
    },
    "stringMinMax": Object {
      "maxLength": 15,
      "minLength": 5,
      "type": "string",
    },
    "tuple": Object {
      "items": Object {
        "oneOf": Array [
          Object {
            "type": "string",
          },
          Object {
            "properties": Object {
              "name": Object {
                "enum": Array [
                  "Rudy",
                ],
                "type": "string",
              },
            },
            "required": Array [
              "name",
            ],
            "type": "object",
          },
          Object {
            "items": Object {
              "oneOf": Array [
                Object {
                  "enum": Array [
                    "blue",
                  ],
                  "type": "string",
                },
                Object {
                  "enum": Array [
                    "red",
                  ],
                  "type": "string",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "type": "array",
    },
    "union": Object {
      "oneOf": Array [
        Object {
          "enum": Array [
            "kek",
          ],
          "type": "string",
        },
        Object {
          "enum": Array [
            "topkek",
          ],
          "type": "string",
        },
      ],
    },
  },
  "required": Array [
    "stringMinMax",
    "numberMinMax",
    "numberMultipleOf",
    "tuple",
    "union",
    "literalString",
    "literalNumber",
    "literalBoolean",
    "array",
    "objectNested",
    "record",
    "recordWithKeys",
    "dateString",
  ],
  "type": "object",
}
`;
